### DISCUSSION TITLE
Make authentication mandatory by default (unless explicitly disabled through a flag)

### DISCUSSION THREAD

**Discussion by drake7707** [Comfy-Org/ComfyUI#5165] [Ideas]
Users of ComfyUI are often unaware of their deployment being exposed, more so when they use a cloud instance, which due to the nature of required hardware specs is very common.

There are over a 1000 exposed ComfyUI installations:
![image](https://github.com/user-attachments/assets/c893f8ae-9f4d-4283-a97d-3b12a9f7b579)
Only 64 of them using ComfyUI-Login:
![image](https://github.com/user-attachments/assets/8c8044e9-3f2e-4378-bca1-f33d03b2fd93)

A lot seem to have ComfyUI-Manager installed which makes it trivially easy to install a custom node that allows shell access to completely take over the instance, install malware, spyware, cryptominers, etc. Anyone could write a script that takes the Shodan results, and do the necessary HTTP requests to have them join a private botnet or whatever, all without the users ever noticing until they get their bill from their cloud provider or their next utilities bill. 

Even without ComfyUI-Manager, it can be used maliciously to host pictures (people really don't want to be a mule for child porn).

Each instance I've visited I've installed ComfyUI-Login so users are at least aware that they their instance is wide open.

I know from reading the opened issues about authentication that it's not really the focus of ComfyUI, nor should it be, but sometimes less tech savvy people need to be protected from themselves or at least made aware that their setup is insecure. Even if authentication is too much work, a simple dismissable popup "You are potentially exposing your ComfyUI instance" when first launched should hopefully give some users pause. People just follow a guide, run a script or copy the --listen 0.0.0.0 without understanding what it means and it can have potential big consequences.

A good similar example was when people setting up raspberry pis that originally came with ssh access on and a default username and password while not realizing they were exposed to the internet and so many pi's got added to a botnet or were used as an attack vector to other devices in the local network. Guides may scream that they should change the password but a lot of people didn't because they didn't understand the implications. The only way to stop that was to force ssh user/password setup. I feel like ComfyUI users falls into the same crowd. Negligence is one thing, but being even unaware that they are being negligent is quite another.

**Comment by ltdrdata** [comment]
That's a good point. At the very least, there should be a warning popup (or visual indicator) when performing operations like `--listen 0.0.0.0`.
@huchenlei ping.

**Comment by wibur0620** [comment]
How can one safely run ComfyUI? Many people don't understand programming and rely on online tutorials for deployment, so they may not know how to ensure a secure environment. If ComfyUI cannot address this directly, creating a detailed guide to teach users how to prevent such issues would be beneficial. What do you think?

**Reply by ltdrdata** [reply]
Basically, you can use it relatively safely by not using `--listen`, so it's not accessible from external sources.

A better method is to run it in a sandbox environment like Docker or a virual environment. In this case, even if damage occurs, you can limit the scope of the risk.

And when installing custom nodes, it is recommended to only use nodes that have been widely used for a long time and have been sufficiently verified.

**Reply by Donzox** [reply]
@ltdrdata If I use ```--listen``` and add localhost to the ports section in the compose.yml? ```- "(localhost):8188:8188"  # Bind to localhost``` 
That will only allow connections from my local machine right? I have a comfy container in wsl and don't plan to connect from any other devices.

**Comment by Xephier102** [comment]
Personally, I worry more about corporations than hackers. You're like, 1000X more likely to have your data unwittingly harvested by a corporation than to have any hacker do anything to you. Cuz the ones making all the security for your computer, are ultimately the ones with easiest access to it. Also, we (effectively) sign so many mile-long legal documents every day just by downloading something and agreeing to use the thing. So, even if we find out they're doing it, chances are, we probably unwittingly agreed to let them prior. 

Naw, I mean, I'll take my chances with hackers, it's the corps that scare the shit outta me. Especially when a corp like Nvidia can make 2 trillion dollars in less than 6 months. Currently Nvidia effectively has enough money to buy Sweden (if it were for sale), and still have more than enough to remain in full business selling video cards. 

Go ahead and ask ChatGPT what 10 of the most evil things are that corporations have done in the past 10 years that have been fully proven. Follow the reading of that with the realization that corporations now have more money than some entire countries. 
And you're afraid of jonny jerkoff in his basement borrowing your rig for crypto..? Admittedly the CP thing may be a bit concerning, but the chances of a hacker even knowing your specific computer exists, are somewhere around that of winning the lottery. 

We hear about it all the time, but that's what the news does, it takes minimal situation events, and makes it look like they're happening everywhere, all the time. 

That said, is it only that --listen thing that opens comfy to the net, or does BFL, M$, Google, or any of the other big corps that make comfy models and/or extensions send your data out(in the background without your knowledge)? Honestly, I worry most about M$, they've literally already been caught using their OS to harvest people's data. That kinda thing is why I use Linux now after being on windows for the past 27ish years.

**Reply by ltdrdata** [reply]
You can review ComfyUI code by yourself. 

Many programmers, besides you, are constantly verifying this.

This is an open source project.

**Reply by Xephier102** [reply]
Oh, I have, or at least I've tried. But I am not actually a programmer. I've only done so because I'm self-sufficient that way, and I'm not afraid to try new things. But, in truth, I don't really know enough to make it practical. For example, I ended up steering clear of unet for some time because I got the sense that it may use online resources, and well, yea, anything that connects to a corporate server can put your data at risk. But, I've learned more since then, and use Unet often, and even got into using flux. I generally prefer to keep my use, local though. 

That said, I only asked because it seemed as if the OP may know enough in that area to give me a definitive answer, or general idea of the security of comfy vs corps, or of any resources to steer clear of (beyond generalized speculation, which I could get from ChatGPT). I mean, why burn all that energy reinventing the wheel, right?

**Reply by drake7707** [reply]
You are literally exposing your machine to become part of a botnet for them to do with as they please. It's the equivalent of leaving your front door wide open, having your address listed as a potential target and then be surprised that at some point all your stuff is gone. There's a reason why front doors have locks.

**Reply by Xephier102** [reply]
> You are literally exposing your machine to become part of a botnet for them to do with as they please. It's the equivalent of leaving your front door wide open, having your address listed as a potential target and then be surprised that at some point all your stuff is gone. There's a reason why front doors have locks.

You're really going to have to be more specific than that. What is it that I've done (or that you think I've done), that would yield those consequences?

**Reply by erstmaldenken** [reply]
@Xephier102 
> the chances of a hacker even knowing your specific computer exists, are somewhere around that of winning the lottery

Ummm... you are aware of tools like https://github.com/robertdavidgraham/masscan ("scanning entire Internet in under 5 minutes")? Nobody needs to know your computer to attack it. It just needs to show up as being vulnerable in a mass scan.

And I'm not afraid of some crypto mining, but ransomware encrypting all data, possibly forever, would be less fun. Probably even less than an evil company selling my personal data.
